benchmark/test_biocypher_query_generation.py

```diff
@@ -140,15 +150,300 @@ def test_end_to_end_query_generation(prompt_engine):
+            f"{prompt_engine.model_name},end-to-end_multi,{calculate_test_score(score),query}\n"
+        )
+
+
+######
+# test hallucination: are all properties available in the KG schema?
+# in selected properties, also in the actual property used in the query
+######
+
+
+def map_entities_to_labels(entity_list):
+    entity_mapping = {}
+    for entity in entity_list:
+        match = re.match(r"(\w+):(\w+)", entity)
+        if match:
+            label, entity_type = match.groups()
+            entity_mapping[label] = entity_type
+
+    return entity_mapping
+
+
+def map_dot_properties_to_labels(property_list):
+    property_mapping = {}
+    for property in property_list:
+        match = re.match(r"(\w+)\.(\w+)", property)
+        if match:
+            label, property_type = match.groups()
+            property_mapping[label] = property_type
+
+    return property_mapping
+
+
+def map_bracket_properties_to_labels(property_list):
+    property_mapping = {}
+    for property in property_list:
+        match = re.search(r"\((\w+):\w+ \{(\w+):", property)
+        if match:
+            label, property_type = match.groups()
+            property_mapping[label] = property_type
+
+    return property_mapping
+
+
+def join_dictionaries(dict1, dict2):
+    result_dict = {}
+    for key in dict1:
+        if key in dict2:
+            result_dict[dict1[key]] = dict2[key]
+
+    return result_dict
+
+
+def get_used_property_from_query(query):
+    property_mapping = dict()
+
+    # first get all properties used in 'dot' format
+
+    property_regex_dot = r"[a-zA-Z]+\.\S+ |[a-zA-Z]+\..+$"
+    used_properties = re.findall(property_regex_dot, query)
+    used_properties = [i.strip() for i in used_properties]
+    # map variable name to properties used
+    property_mapping_add = map_dot_properties_to_labels(used_properties)
+    property_mapping.update(property_mapping_add)
+
+    # get peroperties used in curly brackets
+    if "{" in query:
+        property_regex_bracket = r"\(\w+:\w+ \{\w+: "
+        used_properties = re.findall(property_regex_bracket, query)
+        used_properties = [i.strip() for i in used_properties]
+        # map variable name to properties used
+        property_mapping_add = map_bracket_properties_to_labels(used_properties)
+        property_mapping.update(property_mapping_add)
+
+    # get all entities or relationships involved in the query
+    entity_regex = r"[a-zA-Z]+:\w+"
+    used_entities = re.findall(entity_regex, query)
+    used_entities = [i.strip() for i in used_entities]
+
+    # map variable name to entity or relationship labels
+    entity_mapping = map_entities_to_labels(used_entities)
+
+    # get all the entity and respective properties used in the cypher query
+    used_entity_property = join_dictionaries(entity_mapping, property_mapping)
+
+    return entity_mapping, property_mapping, used_entity_property
+
+
+def test_property_exists(prompt_engine):
+    query = prompt_engine.generate_query(
```

Comment:
I think this should not be the end-to-end process, but rather just the property selection step. This way is less controlled and also costs more.

Reply:
Changed this test to start by selecting properties. The model now tends to invent a new property in the query if it does not pick any properties that it thinks are relevant.